Proposals/20200511-HVPA-design.md

```diff
@@ -0,0 +1,211 @@
+
+### Notes/Constraints/Caveats
+
+#### Different recommendations from HPA and VPA
+In such a case, higher recommendations will be used to calculate the desired replica count. This replica count would be used to calculate desired replica size.
+
+#### Different replica count from total CPU and total memory recommendations
+In such a case, higher replica count would be taken as the desired count. Desired memory and CPU per replica would be calculated by dividing desired total memory and total CPU by the chosen replica count respectively.
+
+#### Pod with multiple containers
+In such a case, all the calculations would be done using the total CPU and total memory allocated to all the containers in the pod for which VPA container resource policy is set to "Auto". The total recommended CPU and memory per replica would then be divided in the same proportion as the recommendations by VPA for those containers.
+
+#### Pod with multiple containers with VPA update resource policy for some containers set to "Off"
+With configuration like this, VPA may not want to scale, however, HPA may want to scale because of load on those containers. In such cases, HVPA controller would scale only horizontally as per HPA recommendations, and will ignore the bucket strategy altogether.
+
+#### User wants to scale only horizontally (TBD)
+<!-- TODO -->
+
+### Risks and Mitigations
+
+#### Base usage of each replica
+Each replica of the application is going to have a base usage. When VPA provides higher recommendations, it includes additional resources required for additional workload. This is because VPA does not consider change in the number of replicas. However, if HVPA ends up scaling horizontally because the aggregated resources lie in a bucket with higher number of replicas, then not all of the additional resources would be available to the actual workload. Some of the resources would be accounted to the base usage of new replicas.
+
+**Example** - Consider a situation where each of the 2 replicas of an application running with 5GB memory has a base memory usage of 2GB because of caching, and the VPA memory recommendations are 6GB. Here the total recommended memory is 12GB. If, according to linearly divided bucket, HVPA needs to scale the applications to 3 replicas, then each replica would be just 4GB so that the total memory allocated to the application is set to 12GB. The memory available after deducting the base usage of all replicas would then be 6GB which is same as before the scaling.
+
+*Mitigation* - HVPA would scale the application to 3 replicas of size 6GB each. The total allocated memory would be 18GB, erring on side of stability rather than the cost.
+
+As a general rule of thumb, HVPA will set the size of each replica to the max of the actual VPA recommendation and the desired resources calculated as per bucket strategy.
```

Comment:
as per previous discussions, we would consider VPA recommendation. Please let me know if you have any new thoughts on why current `requests` should be considered. If we take current `requests`, we would never scale down vertically, isn't it?

Reply:
I thought "the desired resources calculated as per bucket strategy" is calculated based on the VPA recommendation.

Reply:
Yes. it is.
"the desired resources calculated as per bucket strategy" and "actual VPA recommendation" can still be different.
The mitigation, as described, would be to use max of these values for scaling. This is as per our discussion over email.

I am not sure if I have not mentioned any detail in the doc because of which there is this confusion. Please let me know. Thanks.

Reply:
Why would be ever want to use the max of "the desired resources calculated as per bucket strategy" and "actual VPA recommendation"? During scaling up that would always mean "actual VPA recommendation". Unless by "actual VPA recommendation" you mean "current request".

Or am I missing something?

Reply:
Following is a snippet from email discussion:
```
Current Requests: 150m / 1.8Gi with 1 replica
VPA Recommendation: 150m / 2.2Gi
New Requests: 150m / 2.2Gi with 1->2 replicas, because 2.2Gi lands us in the “second” bucket when out-scaling that has 2 replicas; we carry over the vertical scaling as is, which is most conservative.
```

Here "the desired resources calculated as per bucket strategy" would have been 1.1Gi, but we chose to go with 2.2Gi because of the reasons mentioned in the snippet.

Reply:
Sorry, I remembered it as we would apply 1.8Gi rather than 1.1Gi.

Fair enough. But then you might as well just go with max or current requests and VPA recommendation and not bother calculating "the desired resources calculated as per bucket strategy" because it will never have a chance of being selected.

Reply:
"the desired resources calculated as per bucket strategy" becomes useful when scaling down.
As mentioned in the same mail conversation:
```
Current Requests: 150m / 2.2Gi with 3 replicas
VPA Recommendation: 150m / 1.8Gi
New Requests: 225m / 2.7Gi with 3->2 replicas, because 1.8Gi lands us in the “second” bucket when in-scaling that has 2 replicas; we adapt the vertical scaling based on the replica ratio, which is most conservative (3 * 1.8Gi = 5.4 Gi overall memory requests from the PoV of VPA; 5.4 Gi / 2 = 2.7 Gi memory requests per replica that preserves this value; CPU the same, i.e. 3 / 2 * 150m = 225m)
```

Reply:
👍 Then doesn't it make sense to use max of three things? I remember typing this in some email.

1. Current requests
1. Current recommendations
1. the desired resources calculated as per bucket strategy

Reply:
If we use max of these 3, how would an application scale down?

Reply:
👍 Fair enough